archive/sensitivity.lean

```diff
@@ -0,0 +1,434 @@
+/-
+A formalization of Hao Huang's sensitivity theorem: in the hypercube of
+dimension n ≥ 1, if one colors more than half the vertices then at least one
+vertex has at least √n colored neighbors.
+
+A fun summer collaboration by
+Reid Barton, Johan Commelin, Jesse Han, Chris Hughes, Robert Y. Lewis, and Patrick Massot,
+based on Don Knuth's account of the story
+(https://www.cs.stanford.edu/~knuth/papers/huang.pdf),
+using the Lean theorem prover (http://leanprover.github.io/),
+by Leonardo de Moura at Microsoft Research, and his collaborators
+(https://leanprover.github.io/people/),
+and using Lean's user maintained mathematics library
+(https://github.com/leanprover-community/mathlib).
+-/
+import tactic.fin_cases
+import tactic.apply_fun
+import linear_algebra.finite_dimensional
+import linear_algebra.dual
+import analysis.normed_space.basic
+
+-- The next three lines assert we do not want to give a constructive proof,
+-- but rather use classical logic.
+noncomputable theory
+local attribute [instance, priority 1] classical.prop_decidable
+local attribute [instance, priority 0] set.decidable_mem_of_fintype
+
+notation `|`x`|` := abs x
+notation `√` := real.sqrt
+
+open function bool linear_map fintype finite_dimensional dual_pair
+
+/- -------------------------------------------------------------------------\
+|  The hypercube.                                                           |
+\---------------------------------------------------------------------------/
+
+/-
+Notations:
+ℕ denotes natural numbers (including zero).
+fin n = {0, ⋯ , n - 1}.
+bool = {tt, ff}.
+-/
+
+/-- The hypercube in dimension n. -/
+def Q (n : ℕ) := fin n → bool
+
+/-- The projection from Q (n + 1) to Q n forgetting the first value
+(ie. the image of zero). -/
+def π {n : ℕ} : Q (n + 1) → Q n := λ p, p ∘ fin.succ
+
+namespace Q
+-- n will always denote a natural number.
+variable (n : ℕ)
+
+/-- Q n is not empty -/
+instance : inhabited (Q n) := ⟨λ i, tt⟩
+
+/-- Q 0 has a unique element. -/
+instance : unique (Q 0) :=
+⟨⟨λ _, tt⟩, by { intro, ext x, fin_cases x }⟩
+
+/-- Q n is finite. -/
+instance : fintype (Q n) := by delta Q; apply_instance
+
+/-- Q n has 2^n elements. -/
+lemma card : card (Q n) = 2^n :=
+by simp [Q]
+
+/-- The ℝ-module structure on functions from Q n to ℝ (with finite support). -/
+instance coeffs_module (n) : module ℝ (Q n →₀ ℝ) := finsupp.module (Q n) ℝ
```

Comment:
Is this instance used? I don't think we ever really build an isomorphism between these finsupps and the inductively defined `V`.